src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+        return true;
+      } else {
+        String errorMessage = "The teamproject is compulsory when on authorizationMode==teamproject configuration";
+        self.logger.error(errorMessage);
+        WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+          errorMessage);
+        return false;
+      }
+    }
+  }
+
+
+  private boolean checkGen3Authorization(String teamProjectRole, String login) throws Exception {
+    logger.debug("Checking Gen3 Authorization for 'team project'={} and user={} using service={}", teamProjectRole, login, this.authorizationUrl);
+    RestTemplate restTemplate = new RestTemplate();
+    String arboristAuthorizationURL = this.authorizationUrl;
+    String requestBody = String.format("{\"username\": \"%s\"}", login);
+    String jsonResponseString = restTemplate.postForObject(arboristAuthorizationURL, requestBody, String.class);
+
+    JSONObject jsonObject = new JSONObject(jsonResponseString);
+
+    if (!jsonObject.keySet().contains(teamProjectRole)) {
+      logger.warn("User is not authorized to access this team project's data");
+      return false;
+    } else {
+      JSONArray teamProjectAuthorizations = jsonObject.getJSONArray(teamProjectRole);
+      logger.debug("Found authorizations={}", teamProjectAuthorizations);
+      // We expect only one authorization rule per teamproject:
+      if (teamProjectAuthorizations.length() != 1) {
+        logger.error("Only one authorization rule expected for 'teamproject'={}, found={}", teamProjectRole,
+          teamProjectAuthorizations.length());
+        return false;
+      }
+      JSONObject teamProjectAuthorization = teamProjectAuthorizations.getJSONObject(0);
+
+      // check if the authorization contains the right "service" and "method" values:
+      String expectedMethod = "access";
+      String expectedService = "atlas-argo-wrapper-and-cohort-middleware"; // TODO - make the service name configurable?
+      String service = teamProjectAuthorization.getString("service");
+      String method = teamProjectAuthorization.getString("method");
+      logger.debug("Parsed service={} and method={}", service, method);
+      if (!method.equalsIgnoreCase(expectedMethod)) {
+        logger.error("The 'teamproject' authorization method should be '{}', but found '{}'", expectedMethod, method);
+        return false;
+      }
+      logger.debug("Parsed method is as expected");
+      if (!service.equalsIgnoreCase(expectedService)) {
+        logger.error("The 'teamproject' authorization service should be '{}', but found '{}'", expectedService, service);
+        return false;
+      }
+      logger.debug("Parsed service is as expected");
+      return true;
+    }
+  }
+
+  /**
+   * Tries to find the team project information in the given request or as part of the
+   * stored this.authorizer.getCurrentTeamProjectRoleForCurrentUser(). Returns null
+   * if nothing can be found.
```

Comment:
Per my comment about the team project being exposed in the URL which might be copied and reveal Teams to hackers - should we put this in the request header or do a POST and just require that Atlas/WebAPI can only initiate this functionality from e.g., a coded (e.g., JS) http request?

Reply:
I'm not sure I understand the question.
But maybe my previous comments on this topic already clarify the code and the question does not apply anymore?

Reply:
I replied above. I am ok if you close the comment.